### DISCUSSION TITLE
Robust solid state limit switching

### DISCUSSION THREAD

**Discussion by malcolmmp** [phil-barrett/grblHAL-teensy-4.x#110] [Q&A] [answered]
Hi Phil and friends!

I'm a bit stymied right now and I was hoping someone with a little more know-how might be able to help me here.

I've yet to find a set of mechanical limit switches that I find reliable enough for my machine, and inspired by the robustness my LinuxCNC router, I'm aiming to switch over to inductive limit switches.

I've read through the threads and issues I could find on the topic, and while I still have a hard time wrapping my head around sourcing and sinking and how that factors in, I was able to get things working. No problems finding suitable resistors to get the output voltage from 12V down to 5V (or between 4.25-5.25 if I'm remembering a comment on the subject correctly). I experimented with a few different sensors and had best results with an NPN-NC for what it's worth.

My problem comes in when testing the failure modes. In disconnecting different wires, best case is everything fails seen as the limit switch triggering except if either a positive or ground connection is lost, depending on the sensor.

As I understand it, all of the inputs on the bottom of the board have +5V on the optos and are just looking for a path to ground to emit. So grounding the pin registers the input, and no connection or +5V(ish) would be keep the pixies from flowing.

The beauty of sticking an NC mechanical switch in there is that any loose or broken wire will tell the board the limit switch is triggered. I've seen some mention using hall effect sensors, but they're still NPN/PNP transistors, so wouldn't those face the same issue? Is there a way to gain that safety with transistor-based limit switches?

Any help is greatly appreciated!

**Comment by phil-barrett** [comment]
I've been running with NPN/NO proximity sensors in parallel.  

I assume you want to have min and max sensors on a single input.  A single NC prox sensor can be used that way but not multiple in parallel. To gain the fail-alarm feature with multiple NC sensors, you would need to use some logic circuits to combine the signals.

**Comment by malcolmmp** [comment]
Thanks Phil!

I believe the issue came from the resistor I was using as a voltage divider to keep the voltage to the input pin at ~5V. Are the the sensors you're using running directly on 5V?

I ended up having success testing with an opto-coupled DIN mount solid state relay (commonly found as a MRD-060D2). While it drives up my overall cost a bit, I'm a fan of the simplicity of it.

Right now I've just got mechanical switches at my min limits for homing, but the other benefit I see of inductive sensors is being able to use the same sensor for min and max limits, with something as simple as a steel screw head in a t-nut at each end to trigger the limits.

**Reply by phil-barrett** [reply]
A single sensor and 2 targets is pretty cool! The down side is the sensor needs to be on the moving part - not a big deal but adds a little complexity.

**Reply by malcolmmp** [reply]
It's for a pretty large machine with relatively heavy duty gantry. Lots of wires everywhere already, and plenty of room to land the connections. I think for my machine it'll actually simplify the wiring quite a bit.

I've got a large industrial strength cnc router that I'm trying to take a few lessons from regarding construction. The wiring on my grblHal machine has been very rat's-nesty for a couple years now but I'm finally tiding things up. Planning to share some pictures once things are looking presentable!

**Comment by malcolmmp** [comment, marked as answer]
Just a quick follow up for posterity:

I'm attaching a crude diagram of how I've got everything wired with NPN-NC inductive sensors running on 24V connected through opto-coupled solid-state relays (MRD-060D2). Appears to work as described above, where a loss of any wire will be detected in grblHAL as a limit triggered.

Probably better ways of accomplishing all of this, but it's a solution that works well for my needs, so it seems worth sharing.
![Inductive switches with SSRs diagram](https://github.com/user-attachments/assets/a8f56449-2ca0-4b86-aed1-11482060f4ed)

**Comment by rthorntn** [comment]
I just ordered the Teensy board through Tindie!

So I have these switches:

https://www.ebay.com.au/itm/201582829986

PNP NC LJ18A3-8-Z/AY Inductive Proximity Sensor Switch DC6V-36V

How best to wire these in?

Thanks.

**Reply by phil-barrett** [reply]
I've found a lot of them that have a 6V lower limit actually do work at 5V so test for that first.  Otherwise, the Blue goes to Gnd (G terminal) and Black goes to the signal (X, Y, Z,...) terminal.  Wire the brown wire to the +V source. The +V source should have a common ground with the T41U5XBB.

Phil

**Reply by rthorntn** [reply]
Thanks, https://www.grbl.org/single-post/using-pnp-proximity-sensors-with-the-picocnc-and-teensy-boards is saying that I need an opto board for PNP?

**Reply by phil-barrett** [reply]
yes, the board is set up for pull down activation of the optoisolator emitters.  An opto board with NPN output can do that.  Actually, you can use any driver like a transistor.  It needs to pull the signal pin to 0 (Ground).

**Reply by rthorntn** [reply]
I might just buy the NPN version of the switches instead?
